Decouple endpoint ID from addressing

tgraf opened the pull request:
This decouples the endpoint ID from the addressing model and requires plugins such as the CNI plugin to no longer be aware of the endpoint ID at all. Endpoints can be referred to by either container ID or ip addressing information. This allows use of 3rd party IPAM plugins and enables disablement of IPv6.

Depends on: #6079 

*API Consequences:*
* The endpoint ID parameters in PUT /endpoint/{id} becomes meaningless and will be allocated instead
* Clients that blindly assume that the endpoint ID is valid after the API will break. No such clients are known.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6080)
<!-- Reviewable:end -->

ianvernon commented on pkg/endpoint/id/allocator.go:
```diff
@@ -0,0 +1,69 @@
+
+// Release releases an endpoint ID that was previously allocated or reused
+func Release(id uint16) error {
+	if !pool.Insert(idpool.ID(id)) {
```
Why are you checking the result of `pool.Insert` instead of `pool.Release`?

tgraf replied:
Releasing the ID requires to reinsert it into the pool

ianvernon commented on cilium-health/launch/endpoint.go:
```diff
@@ -25,11 +25,9 @@ import (
 	"time"
```
Typo in commit: decopuled --> decoupled

ianvernon commented on pkg/endpointmanager/manager.go:
```diff
@@ -163,6 +178,10 @@ func Remove(ep *endpoint.Endpoint) {
 	delete(endpoints, ep.ID)
 
+	if err := endpointid.Release(ep.ID); err != nil {
+		log.WithError(err).Warning("Unable to relese endpoint ID")
```
release

ianvernon commented on pkg/endpointmanager/manager.go:
```diff
@@ -171,10 +171,14 @@ func Remove(ep *endpoint.Endpoint) {
 		delete(endpointsAux, endpointid.NewID(endpointid.IPv4Prefix, ep.IPv4.String()))
 	}
 
+	if ep.IPv4.IsSet() {
```
This should be `ep.IPv6.IsSet()`

ianvernon commented on pkg/endpoint/id/id.go:
```diff
@@ -71,31 +68,15 @@ func ParseCiliumID(id string) (int64, error) {
+	prefix, id := SplitID(id)
+	switch prefix {
+	case CiliumLocalIdPrefix, CiliumGlobalIdPrefix, ContainerIdPrefix, DockerEndpointPrefix, ContainerNamePrefix:
+	case PodNamePrefix, IPv4Prefix, IPv6Prefix:
```
can't this case collapse in the prior one?

jrajahalme commented on daemon/k8s_watcher_test.go:
```diff
@@ -1156,8 +1167,11 @@ func (ds *DaemonSuite) Test_missingK8sNamespaceV1(c *C) {
 		{
 			name: "endpointmanager doesn't contain any endpoint that is part of that namespace. Should be no-op",
 			setupArgs: func() args {
+				endpointmanager.RemoveAll()
```
Why was this not needed before? Why is this needed now?

tgraf replied:
Not sure why it was not done before, it was not being cleaned up before and relied on the tests never inserting colliding IDs.

jrajahalme commented on pkg/endpoint/id/id.go:
```diff
@@ -45,15 +49,21 @@ func NewID(prefix PrefixType, id string) string {
 // SplitID splits ID into prefix and id. No validation is performed on prefix.
 func SplitID(id string) (PrefixType, string) {
-	if s := strings.Split(id, ":"); len(s) == 2 {
+	if s := strings.SplitN(id, ":", 2); len(s) == 2 {
```
Could use `i := strings.IndexByte(id, ':'); i >= 0` instead, it does not need to allocate memory. Really matters only if we do this a lot.

jrajahalme replied:
return would then be `return id[:i], id[i+1:]`

tgraf replied:
Can be done outside of this PR. Definitely not against it.

jrajahalme reviewed (approved):
We still map an IP address to an ID for `endpointmanager.Lookup()`, how is this decoupling the endpoint ID from addressing?

Other than this, looks good.

aanm commented on pkg/endpointmanager/manager.go:
```diff
@@ -59,7 +59,20 @@ func init() {
+		}
+	} else {
+		id := endpointid.Allocate()
+		if id == uint16(0) {
```
shouldn't it be `id == NoID`?

tgraf replied:
No, it returns an uint16

aanm commented on pkg/endpointmanager/manager.go:
```diff
@@ -163,6 +178,10 @@ func Remove(ep *endpoint.Endpoint) {
 	delete(endpoints, ep.ID)
 
+	if err := endpointid.Release(ep.ID); err != nil {
+		log.WithError(err).Warning("Unable to relese endpoint ID")
```
When would this ever error out?

tgraf replied:
Not sure, but we should not have failure scenarios without a warning of some sort.

tgraf committed:
endpoint: Clean up and document the endpoint/id package

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
endpoint: Check uniqueness of IPv4 and IPv6 address on endpoint creation

So far, it was trusted that the API is fed by IPs reserved by the local IPAM.
Make the API more ressilient by veryfing that the IPs are not used by another
local endpoint.

In order to provide this functinality, cilium endpoint lookup based on the IPv6
address is added.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
endpoint: Introduce endpoint ID allocator

Re-use the idpool to generate random IDs and introduce an endpoint ID
allocation interface.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
endpoint: Allocate endpoint ID independent of IP address

This removes the dependency for an API user to specify a unique endpoint ID
when creating an endpoint and instead allocates the endpoint ID when inserting
the endpoint into the manager.

The endpoint ID is allocated when an endpoint is inserted into the list of
endpoints. When the endpoint ID in the endpoint structure is already set, it is
attempted to be reused to support endpoint restore. For this reason, requests
from plugins creating endpoints with the endpoint ID set are rejected going
forward.

Due to the endpoint ID being allocated on endpoint list insertion, the endpoint
has to be inserted into the list of endpoints before the initial build. This is
behavior change but provides better visibility into the system.

Both the CNI and CMM plugin are adjusted to no longer specify the endpoint ID
in endpoint creation API calls.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
workloads: Correlate workload with endpoint via IP address

This removes the requirement for the workload layer to derive the endpoint ID
via the IP address first.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
addressing: Remove unused code

Signed-off-by: Thomas Graf <thomas@cilium.io>
